pkg/tsdb/postgres/postgres.go

```diff
@@ -42,13 +42,38 @@ func newPostgresQueryEndpoint(datasource *models.DataSource) (tsdb.TsdbQueryEndp
 	return sqleng.NewSqlQueryEndpoint(&config, &queryResultTransformer, newPostgresMacroEngine(timescaledb), logger)
 }
 
-func generateConnectionString(datasource *models.DataSource) string {
+func generateConnectionString(datasource *models.DataSource, logger log.Logger) string {
 	sslmode := datasource.JsonData.Get("sslmode").MustString("verify-full")
+
+	// Always pass SSL mode
+	sslopts := "sslmode=" + url.QueryEscape(sslmode)
+
+	// Attach root certificate if provided
+	if sslrootcert := datasource.JsonData.Get("sslrootcertfile").MustString(""); sslrootcert != "" {
+		logger.Debug("Setting CA certificate", "sslrootcert", sslrootcert)
+		sslopts += "&sslrootcert=" + url.QueryEscape(sslrootcert)
+	}
+
+	// Attach client certificate and key if both are provided
+	sslcert := datasource.JsonData.Get("sslcertfile").MustString("")
+	sslkey := datasource.JsonData.Get("sslkeyfile").MustString("")
+
+	if sslcert != "" && sslkey != "" {
+		logger.Debug("Setting TLS client auth", "sslcert", sslcert, "sslkey", sslkey)
+		sslopts += "&sslcert=" + url.QueryEscape(sslcert) + "&sslkey=" + url.QueryEscape(sslkey)
+
+	} else if (sslcert != "" && sslkey == "") || (sslcert == "" && sslkey != "") {
+		logger.Error("TLS client and certificate must BOTH be specified")
```

Comment:
Please return an error here instead.

Reply:
ahh yeah probably a good call, done in the latest commit.